src/util/Views.h

```diff
@@ -207,9 +207,168 @@ constexpr auto allView(Range&& range) {
+  // be copied (or moved).
+  RvalueView(const RvalueView&) = default;
+  RvalueView& operator=(const RvalueView&) = default;
+  RvalueView(RvalueView&&) = default;
+  RvalueView& operator=(RvalueView&&) = default;
+
+  // Get access to the underlying range.
+  constexpr UnderlyingRange& base() & noexcept { return underlyingRange_; }
+  constexpr const UnderlyingRange& base() const& noexcept {
+    return underlyingRange_;
+  }
+  constexpr UnderlyingRange&& base() && noexcept {
+    return std::move(underlyingRange_);
+  }
+  constexpr const UnderlyingRange&& base() const&& noexcept {
+    return std::move(underlyingRange_);
+  }
+
+  // Begin and end functions implemented using `make_move_iterator`.
+  constexpr auto begin() {
+    return std::make_move_iterator(ql::ranges::begin(underlyingRange_));
+  }
+
+  // TODO<joka921> move_sentinel is C++20, reimplement it.
+  constexpr auto end() {
+    return std::move_sentinel(ql::ranges::end(underlyingRange_));
+  }
+
+  // Const iterators. Note that these will yield `const &&`, which at first
+  // glance seems pointless, but makes the interface more consistent and
+  // generic.
+  CPP_auto_member constexpr auto CPP_fun(begin)()(
+      const  //
+      requires(ql::ranges::range<const UnderlyingRange>)) {
+    return std::make_move_iterator(ql::ranges::begin(underlyingRange_));
+  }
+
+  CPP_auto_member constexpr auto CPP_fun(end)()(
+      const  //
+      requires(ql::ranges::range<const UnderlyingRange>)) {
+    return std::move_sentinel(ql::ranges::end(underlyingRange_));
+  }
+
+  // Size function. Note: The member functions `empty` and `data` are present
+  // via the inheritance from `view_interface` iff they are supported by the
+  // `UnderlyingRange`.
+  CPP_member constexpr auto size()
+      -> CPP_ret(size_t)(requires ql::ranges::sized_range<UnderlyingRange>) {
+    return ql::ranges::size(underlyingRange_);
+  }
+
+  CPP_member constexpr auto size() const -> CPP_ret(size_t)(
+      requires ql::ranges::sized_range<const UnderlyingRange>) {
+    return ql::ranges::size(underlyingRange_);
+  }
+};
+// Deduction guide for `RvalueView`.
+template <typename Range>
+RvalueView(Range&&) -> RvalueView<all_t<Range>>;
+
+// A view that takes another view, but reduces its range category down to
+// `input_range`. In particular, calling `begin` multiple times is disallowed
+// and will throw. A possible application is using this wrapper on the result of
+// a `filter_view` where the values are modified in a way that they don't
+// fulfill the predicate anymore. This is technically undefined behavior, but
+// works in practice if the filter_view is treated as an `input_range`
+CPP_template(typename V)(requires ql::ranges::view<V> CPP_and
+                             ql::ranges::input_range<V>) class ForceInputView
+    : public ql::ranges::view_interface<ForceInputView<V>> {
+ private:
+  V base_;
+  bool beginWasCalled_ = false;
+
+  class Sentinel;
```

Comment:
I don't think you have to forward declare if you write `friend class Xyz` down below, it should do this implicitly.